x-pack/plugins/screenshotting/server/screenshots/event_logger/index.ts

```diff
@@ -0,0 +1,569 @@
+  GET_RENDER_ERRORS = 'get-render-errors',
+  GET_TIMERANGE = 'get-timerange',
+  INJECT_CSS = 'inject-css',
+  REPOSITION = 'position-elements',
+  WAIT_RENDER = 'wait-for-render',
+  WAIT_VISUALIZATIONS = 'wait-for-visualizations',
+  GET_SCREENSHOT = 'get-screenshots',
+  PDF = 'generate-pdf',
+  ADD_IMAGE = 'add-pdf-image',
+  COMPILE = 'compile-pdf',
+}
+
+enum SpanTypes {
+  SETUP = 'setup',
+  READ = 'read',
+  WAIT = 'wait',
+  CORRECT = 'correction',
+  OUTPUT = 'output',
+}
+
+export interface ScreenshottingAction extends LogMeta {
+  event?: {
+    duration?: number; // number of nanoseconds from begin to end of an event
+    provider: typeof PLUGIN_ID;
+  };
+
+  message: string;
+  kibana: {
+    screenshotting: {
+      action: Actions;
+      session_id: string;
+
+      // chromium stats
+      cpu?: number;
+      cpu_percentage?: number;
+      memory?: number;
+      memory_mb?: number;
+
+      // screenshotting stats
+      items_count?: number;
+      pixels?: number;
+      byte_length?: number;
+      element_positions?: number;
+      render_errors?: number;
+
+      // pdf stats
+      byte_length_pdf?: number;
+      pdf_pages?: number;
+    };
+  };
+}
+
+interface ErrorAction {
+  message: string;
+  code?: string;
+  stack_trace?: string;
+  type?: string;
+}
+
+interface GetScreenshotOptions {
+  elementPosition: ElementPosition;
+  byteLength?: number; // byte length of completed completed current screenshot
+}
+
+type SpanEntity =
+  | 'addPdfImage'
+  | 'compilePdf'
+  | 'createPage'
+  | 'getElementPositionData'
+  | 'getNumberOfItems'
+  | 'getRenderErrors'
+  | 'getScreenshot'
+  | 'getTimeRange'
+  | 'injectCss'
+  | 'openUrl'
+  | 'positionElements'
+  | 'waitForRender'
+  | 'waitForVisualization';
+
+type TransactionEntity = 'generatePdf' | 'screenshotting';
+
+type SimpleEvent = Omit<ScreenshottingAction['kibana']['screenshotting'], 'session_id'>;
+
+type LogAdapter = (
+  message: string,
+  event: SimpleEvent,
+  startTime?: Date | undefined
+) => ScreenshottingAction;
+
+function logAdapter(logger: Logger, suffix: 'start' | 'complete', sessionId: string) {
+  const log: LogAdapter = (message, event, startTime) => {
+    let duration: number | undefined;
+    if (startTime != null) {
+      const start = startTime.valueOf();
+      duration = new Date(Date.now()).valueOf() - start.valueOf();
+    }
+
+    const interpretedAction = (event.action + `-${suffix}`) as Actions;
+    const logData: ScreenshottingAction = {
+      message,
+      kibana: {
+        screenshotting: {
+          ...event,
+          action: interpretedAction,
+          session_id: sessionId,
+        },
+      },
+      event: { duration, provider: PLUGIN_ID },
+    };
+    logger.debug(message, logData);
+    return logData;
+  };
+  return log;
+}
+
+/**
+ * A class to use internal state properties to log timing between actions in the screenshotting pipeline
+ */
+export class EventLogger {
```

Comment:
I understand that with this class all of the spans, transactions and their types are encapsulated in one place which makes it more convenient to use, but I am a little bit concerned that this may be hard to maintain over time. Mainly because the structure of this class is coupled to the steps we take in screenshotting.

For example, if we delete or change a step (like rename it) then we will need to update this file too.

I think if we take a factory approach and repurpose your `logAdapter`, creating one for transactions and one for spans it would address the maintainability aspect I was thinking of.  Something like this (excuse the pseudo TS):


```ts

type EndTransaction = (args: {labels: Record<string, string | number >, message: string}) => void;

export class EventLogger {

  createTransaction(args: { message: string, event?: SimpleEvent, id: string }): EndTransaction {
    ...
    this.logger.info(message);
    const transaction = apm.startTransaction(args.action, PLUGIN_ID);
    return (args: {labels: Record<string, string | number >, message: string}) => {
       ...
       Object.entries(labels).forEach(... => transaction.setLabel(...));
       transaction.end();
       this.logger.info(message);
    }
    
    // do something similar for create span
}

```

Then the code that uses this would do:

```ts
import {Actions} from '..'

...
const endScreenshotTransaction = eventLogger.createTransaction({ message: 'starting screenshots', id: Actions.SCREENSHOT })
// do the screenshot
 endScreenshotTransaction();
```

Let me know what you think!

Reply:
Hi @jloleysens I believe I've addressed your feedback and reached the generalization goal you were looking for. The changes were over the course of a few commits:
c8a7b741ad0
94a1f3375d3
ec1ea1552d0

would you mind taking another look?